File: app/src/main/java/com/aero51/githubapp/MainFragment.java
package com.aero51.githubapp;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.paging.PagedList;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.aero51.githubapp.db.model.Repo;

import static com.aero51.githubapp.utils.Constants.DEFAULT_QUERY;
import static com.aero51.githubapp.utils.Constants.LAST_SEARCH_QUERY;
import static com.aero51.githubapp.utils.Constants.LOG;

public class MainFragment extends Fragment {

    private MainViewModel mViewModel;
    private SharedViewModel sharedViewModel;

    private TextView emptyListTextView;
    private RecyclerView recyclerView;
    private MainAdapter adapter;


    public static MainFragment newInstance() {
        return new MainFragment();
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //Get the query to search
         String query = DEFAULT_QUERY;
        if (savedInstanceState != null) {
            query = savedInstanceState.getString(LAST_SEARCH_QUERY, DEFAULT_QUERY);
        }
        mViewModel = new ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(getActivity().getApplication())).get(MainViewModel.class);
        sharedViewModel = new ViewModelProvider(getActivity()).get(SharedViewModel.class);

        //Post the query to be searched
        mViewModel.searchRepo(query);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.main_fragment, container, false);
        emptyListTextView = view.findViewById(R.id.emptyList);
        emptyListTextView.setText(getContext().getResources().getString(R.string.no_results, "\uD83D\uDE13"));
        //Initialize RecyclerView

        recyclerView = view.findViewById(R.id.list);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));


        initAdapter();

        observeSearchQuery();
        observeSortFlags();
        String query = DEFAULT_QUERY;


        return view;
    }




    /**
     * Initializes the Adapter of RecyclerView which is {@link MainAdapter}
     */
    private void initAdapter() {
        adapter = new MainAdapter(getContext());
        recyclerView.setAdapter(adapter);

        //Subscribing to receive the new PagedList Repos
        mViewModel.getRepos().observe(getViewLifecycleOwner(), new Observer<PagedList<Repo>>() {
            @Override
            public void onChanged(PagedList<Repo> repos) {
                if (repos != null) {
                    Log.d(LOG, "initAdapter: Repo List size: " + repos.size());
                    MainFragment.this.showEmptyList(repos.size() == 0);
                    adapter.submitList(repos);
                }
            }
        });
    }

    private void observeSearchQuery(){
        sharedViewModel.getSearchQuery().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(String query) {
                Log.d(LOG, "MainFragment get search query: " + query);
                updateRepoListFromInput(query);
            }
        });
    }
    private void observeSortFlags(){
        sharedViewModel.getSortFlag().observe(getViewLifecycleOwner(), new Observer<Integer>() {
            @Override
            public void onChanged(Integer integer) {
                Log.d(LOG, "MainFragment get sort flag: " + integer);
            }
        });

    }

    /**
     * Shows the Empty view when the list is empty
     *
     * @param show Displays the empty view and hides the list when the boolean is <b>True</b>
     */
    private void showEmptyList(boolean show) {
        if (show) {
            recyclerView.setVisibility(View.GONE);
            emptyListTextView.setVisibility(View.VISIBLE);
        } else {
            recyclerView.setVisibility(View.VISIBLE);
            emptyListTextView.setVisibility(View.GONE);
        }
    }
    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(LAST_SEARCH_QUERY, mViewModel.lastQueryValue());
    }


    private void updateRepoListFromInput(String query) {
        if (!TextUtils.isEmpty(query)) {
            recyclerView.scrollToPosition(0);
            //Posts the query to be searched
            mViewModel.searchRepo(query);
            //Resets the old list
            adapter.submitList(null);
        }
    }
}

File: app/src/main/java/com/aero51/githubapp/MainFragment.kt
package com.aero51.githubapp

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.paging.LoadState
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.aero51.githubapp.databinding.MainFragmentBinding
import com.aero51.githubapp.model.Repo
import com.aero51.githubapp.utils.Constants
import com.aero51.githubapp.utils.SortOrder
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MainFragment : Fragment(R.layout.main_fragment),MainAdapter.OnItemClickListener {
    private val viewModel by viewModels<MainViewModel>()
    private var _binding: MainFragmentBinding? = null
    private val binding get() = _binding!!
    private var searchQuery:String=Constants.DEFAULT_QUERY
    private var sortOrder:SortOrder=SortOrder.BY_STARS
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        _binding = MainFragmentBinding.bind(view)


        val adapter = MainAdapter(this)
        binding.apply {
            recyclerView.setHasFixedSize(true)
            recyclerView.itemAnimator=null
            recyclerView.addItemDecoration(
                    DividerItemDecoration(
                            context,
                            LinearLayoutManager.VERTICAL
                    )
            )
            recyclerView.adapter = adapter.withLoadStateHeaderAndFooter(
                    header = GithubLoadStateAdapter {adapter.retry()},
                    footer = GithubLoadStateAdapter {adapter.retry()}
            )
            buttonRetry.setOnClickListener {
                adapter.retry()
            }
        }
        viewModel.repos.observe(viewLifecycleOwner) {
            adapter.submitData(viewLifecycleOwner.lifecycle, it)
        }
        adapter.addLoadStateListener { loadState ->
            binding.apply {
                progressBar.isVisible = loadState.source.refresh is LoadState.Loading
                recyclerView.isVisible = loadState.source.refresh is LoadState.NotLoading
                buttonRetry.isVisible = loadState.source.refresh is LoadState.Error
                textViewError.isVisible = loadState.source.refresh is LoadState.Error

                // empty view
                if (loadState.source.refresh is LoadState.NotLoading &&
                        loadState.append.endOfPaginationReached &&
                        adapter.itemCount < 1) {
                    recyclerView.isVisible = false
                    textViewEmpty.isVisible = true
                } else {
                    textViewEmpty.isVisible = false
                }
            }
        }
        setHasOptionsMenu(true)
    }



    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)

        inflater.inflate(R.menu.menu_main, menu)

        val searchItem = menu.findItem(R.id.action_search)
        val searchView = searchItem.actionView as SearchView

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {

                if (query != null) {
                    binding.recyclerView.scrollToPosition(0)
                    searchQuery=query
                    viewModel.searchRepos(searchQuery,sortOrder)
                    searchView.clearFocus()
                }
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                return true
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when(item.itemId) {
            R.id.action_sort_by_stars -> {
                sortOrder=SortOrder.BY_STARS
                viewModel.searchRepos(searchQuery,sortOrder)
                true
            }
            R.id.action_sort_by_forks -> {
                sortOrder=SortOrder.BY_FORKS
                viewModel.searchRepos(searchQuery,sortOrder)
                true
            }
            R.id.action_sort_by_updated -> {
                sortOrder=SortOrder.BY_UPDATED
                viewModel.searchRepos(searchQuery,sortOrder)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onItemClick(repo: Repo) {
        val action =MainFragmentDirections.actionMainFragmentToDetailsFragment(repo)
        findNavController().navigate(action)
    }
}